firewood/src/storage/linear/mod.rs

```diff
@@ -96,3 +96,79 @@ impl<S: ReadLinearStore> ReadLinearStore for LinearStore<S> {
         todo!()
     }
 }
+
+#[allow(clippy::unwrap_used, clippy::indexing_slicing)]
+#[cfg(test)]
+pub mod tests {
+    use super::{ReadLinearStore, WriteLinearStore};
+    use std::io::Read;
+    use test_case::test_case;
+
+    #[derive(Debug)]
+    pub struct InMemReadWriteLinearStore {
+        bytes: Vec<u8>,
+    }
+
+    impl InMemReadWriteLinearStore {
+        pub const fn new() -> Self {
+            Self { bytes: vec![] }
+        }
+    }
+
+    impl WriteLinearStore for InMemReadWriteLinearStore {
+        fn write(&mut self, offset: u64, object: &[u8]) -> Result<usize, std::io::Error> {
+            let offset = offset as usize;
+            if offset + object.len() > self.bytes.len() {
+                self.bytes.resize(offset + object.len(), 0);
+            }
+            self.bytes[offset..offset + object.len()].copy_from_slice(object);
+            Ok(object.len())
+        }
+    }
+
+    impl ReadLinearStore for InMemReadWriteLinearStore {
+        fn stream_from(&self, addr: u64) -> Result<impl Read, std::io::Error> {
+            if addr as usize >= self.bytes.len() {
+                return Err(std::io::Error::new(
+                    std::io::ErrorKind::InvalidInput,
+                    "Address out of bounds",
+                ));
+            }
+
+            Ok(&self.bytes[addr as usize..])
+        }
+
+        fn size(&self) -> Result<u64, std::io::Error> {
+            Ok(self.bytes.len() as u64)
+        }
+    }
+
+    #[test_case(&[],0, &[1, 2, 3],&[1, 2, 3]; "write to empty store")]
```

Comment:
change the test so you can just bundle all the writes and then do reads (untested, but should work?)
This lets you test the empty case as well.
```rust
#[test_case(&[(b"123", 0), (b"45", 1)], (0, b"145"))] // overwrite some bytes
#[test_case(&[(b"123", 0), (b"456", 1)], (0, b"1456"))] // extend
#[test_case(&[(b"123", 0), (b"456", 3)], (0, b"123456"))] // only extend

// uncovered cases
#[test_case(&[(b"123", 0), (b"456", 4)], (0, b"123\0456"))] // zero fill
#[test_case(&[], (0, []))] // empty case
#[test_case(&[], (42, []))] // read past eof

// more cases here
fn in_mem_write(mods: &[(u64, &[u8])], expected: (u64, &[u8])) -> Result<(), Error> {
    let mut store = InMemReadWriteLinearStore::default();
    for mod in mods {
        store.write(mod.1, mod.0)?;
    }
    let mut actual = Vec::new();
    store.stream_from(expected.0)?;
    let nread = reader.read_to_end(&mut actual)?;
    assert_eq!(actual.len(), nread);
    assert_eq!(actual, expected.1);
}
```

Reply:
refactored along these lines